zrml/rikiddo/src/types/ema_market_volume.rs

```diff
@@ -0,0 +1,241 @@
+use super::{Timespan, TimestampedVolume, UnixTimestamp};
+use crate::{
+    constants::{EMA_SHORT, SMOOTHING},
+    traits::MarketAverage,
+};
+use frame_support::dispatch::{fmt::Debug, Decode, Encode};
+use substrate_fixed::{
+    traits::{Fixed, FixedSigned, LossyFrom, LossyInto},
+    types::extra::{U24, U32},
+    FixedI32,
+};
+
+#[derive(Clone, Debug, Decode, Encode, Eq, PartialEq)]
+pub struct EmaVolumeConfig<F: Fixed> {
+    pub ema_period: Timespan,
+    pub smoothing: F,
+}
+
+impl<F: FixedSigned + LossyFrom<FixedI32<U24>> + LossyFrom<FixedI32<U32>>> EmaVolumeConfig<F> {
+    pub fn new(ema_period: Timespan, smoothing: F) -> Self {
+        Self { ema_period, smoothing }
+    }
+}
+
+impl<F: FixedSigned + LossyFrom<FixedI32<U24>> + LossyFrom<FixedI32<U32>>> Default
+    for EmaVolumeConfig<F>
+{
+    fn default() -> Self {
+        Self::new(EMA_SHORT, SMOOTHING.lossy_into())
+    }
+}
+
+#[derive(Clone, Debug, Decode, Encode, Eq, PartialEq)]
+pub enum MarketVolumeState {
+    Uninitialized,
+    DataCollectionStarted,
+    DataCollected,
+}
+
+#[derive(Clone, Debug, Decode, Encode, Eq, PartialEq)]
+pub struct EmaMarketVolume<F: Fixed> {
+    pub config: EmaVolumeConfig<F>,
+    pub ema: F,
+    multiplier: F,
+    last_time: UnixTimestamp,
+    state: MarketVolumeState,
+    start_time: UnixTimestamp,
+    volumes_per_period: F,
+}
+
+impl<F: FixedSigned> EmaMarketVolume<F> {
+    pub fn new(config: EmaVolumeConfig<F>) -> Self {
+        Self {
+            config,
+            ema: F::from_num(0),
+            state: MarketVolumeState::Uninitialized,
+            start_time: 0,
+            last_time: 0,
+            volumes_per_period: F::from_num(0),
+            multiplier: F::from_num(0),
+        }
+    }
+
+    fn calculate_ema(&mut self, volume: &TimestampedVolume<F>) -> Result<Option<F>, &'static str> {
+        // Overflow is impossible here (the library ensures that multiplier ∊ [0,1])
+        let volume_times_multiplier = if let Some(res) = volume.volume.checked_mul(self.multiplier)
+        {
+            res
+        } else {
+            return Err("[EmaMarketVolume] Overflow during calculation: volume * multiplier");
+        };
```

Comment:
I decided to include overflow checks, even if they seem impossible to me, for the following reasons:
1) Although I am highly confident that those calculations cannot overflow, I still can be wrong. Therefore I prefer to initially leave them in the pallet, to avoid a node halt due to a panic. If the struct should be kept like this and was tested enough, we could replace them for some minor performance gains if desired.
2) The alternative to use saturated arithmetic does not seem the right way to me, because we would have no way to be informed about misbehavior of the functions except by seeing that the market behaves strange, and we would not have the information where exactly the problem occured.

Reply:
Yes, better play safe than sorry